File: server/models/workflow.go
package models

import (
	"time"
)

type WorkflowTemplateID int
type StepID int
type WorkflowExecID int

type WorkflowTemplate struct {
	ID                WorkflowTemplateID `json:"id" gorm:"primary_key;auto_increment"`
	Label             string             `json:"label"`
	StartStep         StepID             `json:"-"`
	FirstRevisionID   WorkflowTemplateID `json:"first_revision_id"`
	Steps             []*Step            `json:"steps" gorm:"-"`
	StepEditingLocked bool               `json:"step_editing_locked" gorm:"-"`
}

type Step struct {
	ID      StepID             `json:"id" gorm:"primary_key;auto_increment"`
	Label   string             `json:"label"`
	RoomIDs []RoomID           `json:"room_id" gorm:"-"`
	Options map[string][]*Step `json:"options" gorm:"-"`
}

type StepRoom struct {
	StepID StepID `json:"step_id" gorm:"primary_key;auto_increment:false"`
	RoomID RoomID `json:"room_id" gorm:"primary_key;auto_increment:false"`
}

type NextStep struct {
	PrevID        StepID `json:"prev_id" gorm:"primary_key;auto_increment:false"`
	NextID        StepID `json:"next_id" gorm:"primary_key;auto_increment:false"`
	DecisionLabel string `json:"decision_label"`
}

type WorkflowExec struct {
	ID            WorkflowExecID           `json:"id" gorm:"primary_key;auto_increment"`
	Label         string                   `json:"label"`
	Status        WorkflowExecStatus       `json:"status"`
	TemplateID    WorkflowTemplateID       `json:"template_id"`
	TrackerID     TrackerID                `json:"tracker_id"`
	StartedOn     *time.Time               `json:"started_on"`
	CompletedOn   *time.Time               `json:"completed_on"`
	CurrentStepID StepID                   `json:"current_step_id"`
	StepInfos     map[StepID]*ExecStepInfo `json:"step_infos" gorm:"-"`
}

type WorkflowExecStatus int8

const (
	ExecStatusRunning   WorkflowExecStatus = 1
	ExecStatusCompleted WorkflowExecStatus = 2
	ExecStatusCanceled  WorkflowExecStatus = 3
)

func (s WorkflowExecStatus) String() string {
	switch s {
	case ExecStatusRunning:
		return "StatusRunning"
	case ExecStatusCompleted:
		return "StatusCompleted"
	case ExecStatusCanceled:
		return "StatusCanceled"
	}
	return "Unknown Status"
}

type ExecStepInfo struct {
	ExecID      WorkflowExecID `json:"-" gorm:"primary_key;auto_increment:false"`
	StepID      StepID         `json:"-" gorm:"primary_key;auto_increment:false"`
	Decision    string         `json:"decision"`
	StartedOn   *time.Time     `json:"started_on"`
	CompletedOn *time.Time     `json:"completed_on"`
	Skipped     bool           `json:"skipped"`
}

File: app/lib/model/workflow.dart
import 'package:flutter/material.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
import 'package:paper_tracker/widgets/dropdown.dart';

part 'workflow.g.dart';

@JsonSerializable()
class WorkflowTemplate extends DropdownCapable {
  static const IconData = MdiIcons.clipboardFileOutline;
  static const CompletedIconData = MdiIcons.clipboardCheckOutline;

  @JsonKey(name: "id")
  int id;
  @JsonKey(name: "label")
  String label;
  @JsonKey(name: "first_revision_id")
  int firstRevisionID;
  @JsonKey(name: "steps", toJson: _stepsToJson)
  List<WFStep> steps;
  @JsonKey(name: "step_editing_locked")
  bool stepEditingLocked;

  WorkflowTemplate({this.id, this.label, this.firstRevisionID, this.steps, this.stepEditingLocked});

  factory WorkflowTemplate.fromJson(Map<String, dynamic> json) => _$WorkflowTemplateFromJson(json);
  Map<String, dynamic> toJson() => _$WorkflowTemplateToJson(this);
}

dynamic _stepsToJson(List<WFStep> steps) {
  if (steps != null)
    return steps.map((step) => step.toJSON()).toList();
  else
    return null;
}

@JsonSerializable()
class WFStep {
  static const CurrentStepColor = Colors.amber;

  @JsonKey(name: "id")
  int id;
  @JsonKey(name: "label")
  String label;
  @JsonKey(name: "room_ids")
  List<int> roomIDs;
  @JsonKey(name: "options", includeIfNull: false, toJson: _optionsToJson)
  Map<String, List<WFStep>> options;

  WFStep({this.id, this.label, this.roomIDs, this.options});

  factory WFStep.fromJson(Map<String, dynamic> json) => _$WFStepFromJson(json);
  Map<String, dynamic> toJSON() => _$WFStepToJson(this);
}

dynamic _optionsToJson(Map<String, List<WFStep>> options) {
  if (options != null)
    return options.map((decision, steps) => MapEntry(decision, _stepsToJson(steps)));
  else
    return null;
}

@JsonSerializable()
class WorkflowExec {
  static const IconData = MdiIcons.clipboardTextPlayOutline;
  static const CompletedColor = Colors.teal;

  @JsonKey(name: "id")
  int id;
  @JsonKey(name: "label")
  String label;
  @JsonKey(name: "template_id")
  int templateID;
  @JsonKey(name: "tracker_id")
  int trackerID;
  @JsonKey(name: "status")
  WorkflowExecStatus status;
  @JsonKey(name: "started_on")
  DateTime startedOn;
  @JsonKey(name: "completed_on")
  DateTime completedOn;
  @JsonKey(name: "current_step_id")
  int currentStepID;
  @JsonKey(name: "step_infos", toJson: _stepInfosToJSON)
  Map<int, ExecStepInfo> stepInfos;

  WorkflowExec(
      {this.id,
      this.label,
      this.templateID,
      this.trackerID,
      this.status,
      this.startedOn,
      this.completedOn,
      this.currentStepID,
      this.stepInfos});

  factory WorkflowExec.fromJson(Map<String, dynamic> json) => _$WorkflowExecFromJson(json);
  Map<String, dynamic> toJSON() => _$WorkflowExecToJson(this);
}

dynamic _stepInfosToJSON(Map<int, ExecStepInfo> stepInfos) {
  return stepInfos?.map((k, e) => MapEntry(k.toString(), e.toJSON()));
}

enum WorkflowExecStatus {
  @JsonValue(1)
  Running,
  @JsonValue(2)
  Finished,
  @JsonValue(3)
  Cancelled,
}

extension WorkflowExecStatusExtension on WorkflowExecStatus {
  String get label {
    return this.toString().substring(this.toString().indexOf('.') + 1);
  }

  IconData get icon {
    switch (this) {
      case WorkflowExecStatus.Running:
        return Icons.play_circle_filled;
      case WorkflowExecStatus.Finished:
        return Icons.done;
      case WorkflowExecStatus.Cancelled:
        return Icons.cancel;
    }
    return Icons.adb;
  }
}

@JsonSerializable()
class ExecStepInfo {
  @JsonKey(name: "decision")
  String decision;
  @JsonKey(name: "started_on")
  DateTime startedOn;
  @JsonKey(name: "completed_on")
  DateTime completedOn;
  @JsonKey(name: "skipped")
  bool skipped;

  ExecStepInfo({this.decision, this.startedOn, this.completedOn, this.skipped});

  factory ExecStepInfo.fromJson(Map<String, dynamic> json) => _$ExecStepInfoFromJson(json);
  Map<String, dynamic> toJSON() => _$ExecStepInfoToJson(this);
}
